src/busctl/busctl.c

```diff
@@ -2261,6 +2261,93 @@ static int get_property(int argc, char **argv, void *userdata) {
         return 0;
 }
 
+static int on_bus_signal(sd_bus_message *msg, void *userdata, sd_bus_error *ret_error) {
+        bool *fini = userdata;
+        int r;
+
+        *fini = true;
+
+        r = sd_bus_message_is_empty(msg);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        if (r == 0 && !arg_quiet) {
+
+                if (!FLAGS_SET(arg_json_format_flags, SD_JSON_FORMAT_OFF)) {
+                        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+                        if (arg_json_format_flags & (SD_JSON_FORMAT_PRETTY|SD_JSON_FORMAT_PRETTY_AUTO))
+                                pager_open(arg_pager_flags);
+
+                        r = json_transform_message(msg, &v);
+                        if (r < 0)
+                                return r;
+
+                        sd_json_variant_dump(v, arg_json_format_flags, NULL, NULL);
+
+                } else if (arg_verbose) {
+                        pager_open(arg_pager_flags);
+
+                        r = sd_bus_message_dump(msg, stdout, 0);
+                        if (r < 0)
+                                return r;
+                } else {
+
+                        fputs(sd_bus_message_get_signature(msg, true), stdout);
+                        fputc(' ', stdout);
+
+                        r = format_cmdline(msg, stdout, false);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+
+                        fputc('\n', stdout);
+                }
+        }
+
+        return 0;
+}
+
+static int wait_signal(int argc, char **argv, void *userdata) {
+        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        bool fini = false;
+        int r;
+
+        const char *sender = argv[1];
+        const char *path = argv[2];
+        const char *interface= argv[3];
+        const char *member = argv[4];
+
+        r = acquire_bus(false, &bus);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_match_signal(bus, NULL, sender, path, interface, member, on_bus_signal, &fini);
+        if (r < 0)
+                return log_error_errno(r, "Failed to match signal %s on interface %s: %m", member, interface);
+
```

Comment:
i think it would make sense to add `sd_notify("READY=1")` here, so that a caller can use that to know when the signal matcher has been successfully installed. Compare with how monitor() does this, which is a very similar case.